src/sorted_list/mod.rs

```diff
@@ -0,0 +1,252 @@
+    }
+
+    /// Returns a reference to the last (maximum) value in the list.
+    pub fn last(&mut self) -> Option<&T> {
+        self.lists.last().and_then(|x| x.last())
+    }
+
+    pub fn last_mut(&mut self) -> Option<&mut T> {
+        self.lists.last_mut().and_then(|x| x.last_mut())
+    }
+
+    pub fn pop_first(&mut self) -> Option<T> {
+        if self.len() == 0 {
+            None
+        } else {
+            self.len -= 1;
+            let rv = Some(self.lists[0].remove(0));
+            self.contract(0);
+            rv
+        }
+    }
+
+    pub fn pop_last(&mut self) -> Option<T> {
+        if let Some(rv) = self.lists.last_mut().and_then(|l| l.pop()) {
+            self.len -= 1;
+            let len = self.len;
+            self.contract(len);
+            Some(rv)
+        } else {
+            None
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    pub fn iter(&self) -> Iter<T> {
+        let mut outer = self.lists.iter();
+        let inner = outer.next().unwrap().iter();
+        Iter { outer, inner }
+    }
+}
+
+impl<T: Ord> Index<usize> for SortedList<T> {
+    type Output = T;
+
+    fn index(&self, i: usize) -> &T {
+        let mut i = i;
+        for list in &self.lists {
+            if list.len() > i {
+                return &list[i];
+            } else {
+                i = i - list.len();
+            }
+        }
+        panic!("element greater than list size");
+    }
+}
+
+impl<T: Ord> IndexMut<usize> for SortedList<T> {
+    fn index_mut(&mut self, i: usize) -> &mut T {
+        let mut i = i;
+        for list in &mut self.lists {
+            if list.len() > i {
+                return &mut list[i];
+            } else {
+                i = i - list.len();
+            }
+        }
+        panic!("element greater than list size");
+    }
+}
+
+pub struct Iter<'a, T: 'a> {
+    outer: std::slice::Iter<'a, Vec<T>>,
+    inner: std::slice::Iter<'a, T>,
+}
+
+impl<'a, T: Ord> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next().or_else(|| {
+            self.outer.next().and_then(|x| {
+                self.inner = x.into_iter();
+                self.next()
+            })
+        })
+    }
+}
+
+pub struct IntoIter<T> {
+    outer: std::vec::IntoIter<Vec<T>>,
+    inner: std::vec::IntoIter<T>,
+}
+
+impl<T: Ord> Iterator for IntoIter<T> {
+    type Item = T;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next().or_else(|| {
+            self.outer.next().and_then(|x| {
+                self.inner = x.into_iter();
+                self.next()
+            })
+        })
+    }
+    // Size hint doesn't seem super helpful
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        match (self.outer.size_hint(), self.inner.size_hint()) {
+            ((o_min, Some(o_max)), (i_min, Some(i_max))) => (o_min + i_min, Some(o_max + i_max)),
```

Comment:
This logic looks funky now that I look at it. We should revisit assumptions about the number of elements it is possible to have in a list.
We should fix this... maybe after this PR lands.